scripts/remove-unregistered-users.ts

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env node -r esm -r ts-node/register
+
+import { resolve } from "path";
+
+import admin from "firebase-admin";
+
+import {
+  checkFileExists,
+  initFirebaseAdminApp,
+  makeScriptUsage,
+  parseCredentialFile,
+} from "./lib/helpers";
+
+const usage = makeScriptUsage({
+  description: `Bulk remove of user accounts that exist in the auth but are not in the firestore users collection`,
+  usageParams: `CREDENTIAL_PATH`,
+  exampleParams: `fooAccountKey.json`,
+});
+
+const [credentialPath] = process.argv.slice(2);
+if (!credentialPath) {
+  usage();
+}
+
+if (!checkFileExists(credentialPath)) {
+  console.error("Credential file path does not exists:", credentialPath);
+  process.exit(1);
+}
+
+const { project_id: projectId } = parseCredentialFile(credentialPath);
+
+if (!projectId) {
+  console.error("Credential file has no project_id:", credentialPath);
+  process.exit(1);
+}
+
+initFirebaseAdminApp(projectId, {
+  credentialPath: credentialPath
+    ? resolve(__dirname, credentialPath)
+    : undefined,
+});
+
+// 1000 is the maximum number of users firestore can fetch. check .listUsers()'s documentation for more information.
+// It is also 1000 by default but I decided to set it explicitly to avoid confusion.
+const MAX_USERS = 1000;
+
+const getAuthUsers = async (maxUsers: number, pageToken?: string) =>
+  await admin.auth().listUsers(maxUsers, pageToken);
+
+(async () => {
+  const totalAuthUsers: admin.auth.UserRecord[] = [];
+
+  console.log("Fetching firestore auth users...");
+  const { users: authUsers, pageToken } = await getAuthUsers(MAX_USERS);
+  totalAuthUsers.push(...authUsers);
+
+  // @debt This is going to make another request with offset if there are the maximum users returned by the previous request. This should be using recursion.
```

Comment:
[NB] Why should it be using recursion? To avoid mutation/etc?

Reply:
Because it will fetch only 2k users and nothing more. We need to add recursion and every time it fetches 1000 users it should trigger again. Where right now, it happens only once.

Reply:
As I can see from the file diff, the recursion was removed and now we only fetch 2k users at most? Doesn't seem right to me.

Reply:
Ah, makes sense. Instead of recursion this seems to just be handled with loops and modifying result variables.

While I’m not so much a fan of that pattern in general, and recursion would probably allow us to avoid it, I also don’t think the solution here would need to inherently involve recursion.

Could just be something like: while (stillMoreToGet) results.push(…getMore())

Reply:
@mike-lvov the recursion was never there and it only ever fetches 2k users. Which I missed, and what @denisdimitrov explained above.